File: src/engine/vec2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine
{
    public class vec2
    {
        public float x, y;
        public vec2(float x, float y)
        {
            this.x = x;
            this.y = y;
        }

        public static vec2 Zero
        {
            get
            {
                return new vec2(0, 0);
            }
        }

        public override string ToString()
        {
            return string.Format("({0} {1})", x, y);
        }

        public static vec2 operator +(vec2 lhs, vec2 rhs)
        {
            return new vec2(lhs.x + rhs.x, lhs.y + rhs.y);
        }
        public static vec2 operator -(vec2 lhs, vec2 rhs)
        {
            return new vec2(lhs.x - rhs.x, lhs.y - rhs.y);
        }
        public static vec2 operator *(vec2 lhs, float rhs)
        {
            return lhs.scale(rhs);
        }

        private vec2 scale(float s)
        {
            return new vec2(x*s, y*s);
        }

        public static vec2 FromTo(vec2 from, vec2 to)
        {
            return to - from;
        }

        public static vec2 Curve(vec2 target, vec2 old, float smoothing)
        {
            return new vec2(math1.Curve(target.x, old.x, smoothing), math1.Curve(target.y, old.y, smoothing));
        }
    }
}

File: src/engine/vec2.cc
#include "engine/vec2.h"

#include "engine/math.h"
#include "fmt/core.h"

namespace simple
{
    vec2::vec2(float xx, float yy)
        : x(xx)
        , y(yy)
    {
    }

    vec2 vec2::Zero()
    {
        return vec2(0, 0);
    }

    std::string vec2::ToString() const
    {
        return fmt::format("({} {})", x, y);
    }

    vec2 vec2::GetScaled(float s) const
    {
        return vec2(x * s, y * s);
    }

    vec2 vec2::FromTo(const vec2& from, const vec2& to)
    {
        return to - from;
    }

    vec2 vec2::Curve(const vec2& target, const vec2& old, float smoothing)
    {
        return vec2(math::Curve(target.x, old.x, smoothing), math::Curve(target.y, old.y, smoothing));
    }

    vec2 operator+(const vec2& lhs, const vec2& rhs)
    {
        return vec2(lhs.x + rhs.x, lhs.y + rhs.y);
    }

    vec2 operator-(const vec2& lhs, const vec2& rhs)
    {
        return vec2(lhs.x - rhs.x, lhs.y - rhs.y);
    }

    vec2 operator*(const vec2& lhs, float rhs)
    {
        return lhs.GetScaled(rhs);
    }
}
